### PR TITLE
Use filter query when describing instances

### PR DESCRIPTION
**Pull request by res0nance**
- Get the API to do the filtering instead of doing it ourselves

### CODE REVIEW

File: src/main/java/hudson/plugins/ec2/EC2Cloud.java [hunk 1]
```diff
@@ -386,42 +386,29 @@ private int countCurrentEC2Slaves(SlaveTemplate template) throws AmazonClientExc
         Set<String> instanceIds = new HashSet<String>();
         String description = template != null ? template.description : null;
 
-        //FIXME convert to a filter query
-        for (Reservation r : connect().describeInstances().getReservations()) {
+        List<Filter> filters = getGenericFilters(jenkinsServerUrl, template);
+        filters.add(new Filter("instance-state-name").withValues("running", "pending", "stopping"));
+        if (template != null) {
+            filters.add(new Filter("image-id").withValues(template.getAmi()));
```
Comment (res0nance) [inline src/main/java/hudson/plugins/ec2/EC2Cloud.java]: Slightly worried that this may cause issues if someone changes the `ami_id`

**Comment by res0nance** [comment]
@jvz I think I'm done with this just turning it into the equivalent call letting the aws api deal with as much of the filtering as possible

**Comment by jvz** [comment]
Can you either add a test or describe how to manually test this?

**Comment by res0nance** [comment]
Behaviour doesn't change it has the exact behaviour as before as for automated testing we need a way to mock the aws api for that

**Comment by res0nance** [comment]
@thoulen Can you take a look at this? This should help reduce the amount of data we pull from aws api and offload some of the filtering logic back to aws.

**Comment by thoulen** [comment]
I am testing the patch

**Comment by res0nance** [comment]
Added a fix to take care of a possible miscount when nexttoken is available

**Comment by res0nance** [comment]
This PR also includes a bugfix where instance counting will be wrong if there are more than 1000 instances because we ignore the next token

**Comment by res0nance** [comment]
Tested this and my instance caps are still properly respected so I'm merging it.

**Pull request merged**